# Strings in Python

A string is a **sequence** of characters. Are used to represent a piece of text.

```python
s = "this is a string"
s = """this is
a text
on several
lines"""
s = 'this is another test'
s = "John's dog"
s = "This is an invalid string'

```

- strings can contain escape sequences in order to contain special characters
- an escape sequence is marked by ``\`` (backslash) followed by a letter (``\n`` new line, ``\t`` tab, ``\\`` for ``\``)

# Reading an integer

We can use ``input`` to read an integer, but we need to convert the output to ``int``

```python
stringVariable = input("Enter an integer:")
numericVariable = int(stringVariable)
```
- we can convert a string to an integer using ``int(string)``. If ``string`` is not a valid integer an exception is thrown
- ``float(string)`` is used to convert a string to float
- ``str(numeric_value)`` converts a numeric value to string

Read more about this on pages 42-43 of *Python Programming for the Absolute Beginner* by Michael Dawson

In [1]:
# convert a string to integer
int("12345")

12345

In [2]:
# convert a string to float
float("12.34")

12.34

In [3]:
# trying to convert invalid values
int("123s")

ValueError: invalid literal for int() with base 10: '123s'

In [4]:
int("12.34")

ValueError: invalid literal for int() with base 10: '12.34'

# String methods

- there are large number of methods that implement various important string processing 
   - for changing capitalisation: ``capitalize``, ``casefold``, ``lower``, ``swapcase``, ``title``, ``upper``
   - alignment: ``center``, ``ljust``, ``rjust``, ``zfill``
   - counting and finding: ``count``, ``endswith``, ``find``, ``index``, ``rfind``, ``rindex``, ``startswith``
   - transformation: ``encode``, ``expandtabs``, ``format``, ``join``, ``lstrip``, ``partition``, ``replace``, ``rpartition``, ``rsplit``, ``rstrip``, ``split``, ``splitlines``, ``strip``, ``translate``
   - testing: ``isalnum``, ``isalpha``, ``isdecimal``, ``isdigit``, ``isidentifier``, ``islower``, ``isnumeric``, ``isprintable``, ``isspace``, ``istitle``, ``isupper``
   
- in many cases the names are quite suggestive to guess what they mean
- full details are available at <a href="https://docs.python.org/3.7/library/stdtypes.html#string-methods" target="_blank">https://docs.python.org/3.7/library/stdtypes.html#string-methods</a>

# Example of how *find* is used

In [5]:
help(str.find)

Help on method_descriptor:

find(...)
    S.find(sub[, start[, end]]) -> int
    
    Return the lowest index in S where substring sub is found,
    such that sub is contained within S[start:end].  Optional
    arguments start and end are interpreted as in slice notation.
    
    Return -1 on failure.



In [6]:
line = "this is a test line"
line.find("test")

10

**Note**: counting in strings (and other sequences) starts at **0**

# Other operations with strings

- These are shared with other types of sequences (e.g. lists, tuples)

- ``+`` concatenations of strings (e.g. ``"cat" + "s"``)
- ``len(string)`` returns the length of the string (``len("word")``)
- ``*`` repetition of the sequence (e.g. ``"a"*2``)
- ``in`` operator tests whether an element is in a sequence (e.g. ``"1" in "0123"``)

# Exercise

- look at various methods related to strings. Understand what they are doing. Are any of them potentially useful for your work?
- learn more about splitting, concatenating and joining string: <a href="https://realpython.com/python-string-split-concatenate-join/" target="_blank">https://realpython.com/python-string-split-concatenate-join/</a>

# Further readings

- Try examples from <a href="https://en.wikibooks.org/wiki/Non-Programmer%27s_Tutorial_for_Python_3/Who_Goes_There%3F" target="_blank">https://en.wikibooks.org/wiki/Non-Programmer%27s_Tutorial_for_Python_3/Who_Goes_There%3F</a>

- Read Chapter 2 of *Python Programming for the Absolute Beginner* by Michael Dawson

- A comprehensive discussion about strings: <a href="https://realpython.com/python-strings/" target="_blank">https://realpython.com/python-strings/</a>

